src/test/kotlin/com/zepben/cimbend/get/CustomerConsumerClientTest.kt

```diff
@@ -152,10 +153,46 @@ internal class CustomerConsumerClientTest {
 
         consumerClient.removeErrorHandler(onErrorHandler)
 
-        val result = consumerClient.getIdentifiedObjects(service, mRIDs)
+        val objects = consumerClient.getIdentifiedObjects(service, mRIDs)
 
         verify(stub).getIdentifiedObjects(GetIdentifiedObjectsRequest.newBuilder().addAllMrids(mRIDs).build())
-        validateFailure(onErrorHandler, result, expectedEx, false)
+        validateFailure(onErrorHandler, objects, expectedEx, false)
+    }
+
+    @Test
+    internal fun `returns failed mRID when duplicate mRID exists in the service`() {
+        val mRIDs = listOf("id1", "id1", "id2")
+        val response1 = createResponse(CustomerIdentifiedObject.newBuilder(), CustomerIdentifiedObject.Builder::getCustomerBuilder, mRIDs[0])
+        val response2 = createResponse(CustomerIdentifiedObject.newBuilder(), CustomerIdentifiedObject.Builder::getCustomerBuilder, mRIDs[1])
+        val response3 = createResponse(CustomerIdentifiedObject.newBuilder(), CustomerIdentifiedObject.Builder::getCustomerAgreementBuilder, mRIDs[2])
+
+        doReturn(listOf(response1, response2, response3).iterator()).`when`(stub).getIdentifiedObjects(any())
+
+        val objects = consumerClient.getIdentifiedObjects(service, mRIDs)
+
+        assertThat(objects.wasSuccessful, equalTo(true))
+        assertThat(objects.value.result.size, equalTo(2))
+        assertThat(objects.value.result[mRIDs[0]], instanceOf(Customer::class.java))
+        assertThat(objects.value.result[mRIDs[2]], instanceOf(CustomerAgreement::class.java))
+        assertThat(objects.value.failed, Matchers.contains(mRIDs[0]))
+
+        verify(stub).getIdentifiedObjects(GetIdentifiedObjectsRequest.newBuilder().addAllMrids(listOf("id1", "id2")).build())
+        clearInvocations(stub)
+    }
```

Comment:
This test (and the ones in the other similar classes) are very misleading and confusing, took me a while to figure out how it was returning the null (forcing a failure) you are checking for.

The mRID's that you request have nothing to do with the results sent back, they are mocked out with an `any()` matcher.

What this is really testing is that if you get a result that you have already processed that it creates an error report for the duplicate, which you can do just by returning the same result twice in the stream and requesting any mRID that causes the request to fire. e.g. `doReturn(listOf(response1, response1).iterator()).`when`(stub).getIdentifiedObjects(any())`

The current form of passing in multiple mRID's that are the same leads to thinking you are making multiple requests, which we are not due to collecting to a set.

Probably needs to be reworked and have a comment added to the test indicating that we are checking for handling of duplicate responses, not duplicate requests, for safe guarding against a bad server response/race conditions etc.

Reply:
Think I understand what you mean - made some changes, lmk what you think.